src/main/java/delight/nashornsandbox/internal/NashornSandboxImpl.java

```diff
@@ -96,40 +101,73 @@ public NashornSandboxImpl(String... params) {
 	}
 
     private void assertScriptEngine() {
-        assertScriptEngine(scriptEngine.getBindings(ScriptContext.ENGINE_SCOPE));
-    }
-
-	private void assertScriptEngine(Bindings bindings) {
-		try {
-            scriptEngine.eval(preScript(), sanitizeBindings(bindings));
-		} catch (final Exception e) {
-			throw new RuntimeException(e);
-		}
-	}
-
-	private String preScript() {
-        final StringBuilder sb = new StringBuilder();
-        if (!allowExitFunctions) {
-            sb.append("var quit=function(){};var exit=function(){};");
+        try {
+            if (!engineAsserted) {
+                produceSecureBindings();
+            }
+            else if (!engineBindingsSafe()) {
+                resetEngineBindings();
+            }
         }
-        if (!allowPrintFunctions) {
-            sb.append("var print=function(){};var echo = function(){};");
+        catch (final Exception e) {
+            throw new RuntimeException(e);
         }
-        if (!allowReadFunctions) {
-            sb.append("var readFully=function(){};").append("var readLine=function(){};");
+    }
+
+    // This will detect whether the current engine bindings match the cached protected bindings
+    private boolean engineBindingsSafe() {
```

Comment:
Could we rename to `engineBindingsUnchanged`?

Reply:
I have renamed it as suggested, it is much clearer this way.